File: kotlin/src/test/java/kotlin/PublishedApiTest.java
package kotlin;

import org.junit.Test;
import params.Param;
import params.ParamsKt;

import java.lang.reflect.Method;

import static org.hamcrest.Matchers.equalTo;
import static org.junit.Assert.assertThat;

public class PublishedApiTest {
    @Test
    public void canNotBeAccessSyntheticPublishedApiDirectly() throws Throwable {
        @SuppressWarnings("JavaReflectionMemberAccess")
        Method published = ParamsKt.class.getDeclaredMethod("singleton", Object.class);

        @SuppressWarnings("unchecked")
        Param<String> param = (Param<String>) published.invoke(null, "foo");

        assertThat(param.map(it -> it), equalTo("foo"));
    }
}

File: kotlin/src/test/kotlin/test/PublishedApiTest.kt
package test

import com.natpryce.hamkrest.assertion.assert
import com.natpryce.hamkrest.equalTo
import org.junit.Test
import params.paramOf

class PublishedApiTest {

    @Test
    fun `build param`() {
        val value = paramOf("foo") { map { "$it-bar" } }

        assert.that(value, equalTo("foo-bar"))
    }
}